### PR TITLE
Feat: update project status

### PR DESCRIPTION
**Pull request by datphamcode295**
#### What's this PR do?

- [x] API update project status

#### What are the relevant Git tickets?
https://www.notion.so/API-to-update-project-status-f98e5f7851614e03aaae5d2e073c95c2

#### Screenshots (if appropriate)
image

#### Request
```
curl 'http://localhost:8200/api/v1/projects/8dc3be2e-19a4-4942-8a79-56db391a0b15/status'
-H 'accept: /'
-H 'accept-language: vi,en-US;q=0.9,en;q=0.8'
-H 'content-type: application/json'
-H 'origin: http://localhost:3000/'
-H 'referer: http://localhost:3000/'
-H 'sec-ch-ua: "Chromium";v="106", "Google Chrome";v="106", "Not;A=Brand";v="99"'
-H 'sec-ch-ua-mobile: ?0'
-H 'sec-ch-ua-platform: "macOS"'
-H 'sec-fetch-dest: empty'
-H 'sec-fetch-mode: cors'
-H 'sec-fetch-site: cross-site'
-H 'user-agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/106.0.0.0 Safari/537.36'
-H 'authorization:Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE2OTkzMjExNDIsImlkIjoiMjY1NTgzMmUtZjAwOS00YjczLWE1MzUtNjRjM2EyMmU1NThmIiwiYXZhdGFyIjoiaHR0cHM6Ly9zMy1hcC1zb3V0aGVhc3QtMS5hbWF6b25hd3MuY29tL2ZvcnRyZXNzLWltYWdlcy81MTUzNTc0Njk1NjYzOTU1OTQ0LnBuZyIsImVtYWlsIjoidGhhbmhAZC5mb3VuZGF0aW9uIiwicGVybWlzc2lvbnMiOlsiZW1wbG95ZWVzLnJlYWQiXSwidXNlcl9pbmZvIjpudWxsfQ.GENGPEucSUrILN6tHDKxLMtj0M0REVMUPC7-XhDMpGM'
--data-raw '{"status":"active"}'
--compressed
```

#### Response
```
{
    "data": {
        "id": "8dc3be2e-19a4-4942-8a79-56db391a0b15",
        "createdAt": "2022-11-11T18:06:56.362902Z",
        "updatedAt": "2022-11-14T18:11:22.113457Z",
        "name": "Fortress",
        "type": "dwarves",
        "status": "active",
        "startDate": "2022-11-01T00:00:00Z",
        "endDate": null
    }
}
```

### CODE REVIEW

File: pkg/view/project.go [hunk 1]
```diff
@@ -38,6 +38,73 @@ type ProjectHead struct {
 	Avatar      string `json:"avatar"`
 }
 
+type UpdateProjectStatusResponse struct {
+	Data ProjectData `json:"data"`
+}
+
+func ToUpdateProjectStatusResponse(p *model.Project) ProjectData {
```
Comment (namnhce) [inline pkg/view/project.go:45]: Can you return to another data structure? `ProjectData` here, should be used for listing. So I will update the name of this struct later. Btw, you can create a new one and return the same thing with model.Project

File: pkg/handler/project/testdata/update_project_status/200.json [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+        "name": "Fortress",
+        "type": "dwarves",
+        "status": "active",
+        "startDate": "2022-11-01T00:00:00Z",
+        "endDate": null,
+        "members": [],
+        "technicalLeads": [],
+        "accountManager": null,
+        "salePerson": null,
+        "deliveryManager": null
```
Comment (namnhce) [inline pkg/handler/project/testdata/update_project_status/200.json:15]: we dont want to see the data from line 11 -> 15

**Review by namnhce — CHANGES_REQUESTED** [review]

File: pkg/handler/project/testdata/update_project_status/200.json (hunk 2 above)
Comment (datphamcode295) [reply pkg/handler/project/testdata/update_project_status/200.json]: I create new struct to return it

File: pkg/handler/project/project.go [hunk 3]
```diff
@@ -78,3 +79,59 @@ func (h *handler) List(c *gin.Context) {
...
+		}
+	}
+
+	// 1.2 prepare the logger
+	// TODO: can we move this to middleware ?
+	l := h.logger.Fields(logger.Fields{
+		"handler": "project",
+		"method":  "UpdateProjectStatus",
+		"params":  projectID,
+	})
```
Comment (nguyennh4522) [inline pkg/handler/project/project.go:119]: add `body` to logger

File: pkg/handler/project/project.go [hunk 4]
```diff
@@ -78,3 +79,59 @@ func (h *handler) List(c *gin.Context) {
...
+// @Success 200 {object} view.UpdateProjectStatusResponse
+// @Failure 400 {object} view.ErrorResponse
+// @Failure 404 {object} view.ErrorResponse
+// @Failure 500 {object} view.ErrorResponse
+// @Router /projects/{id}/status [post]
+func (h *handler) UpdateProjectStatus(c *gin.Context) {
+	// 1. parse id from uri, validate id
+	projectID := c.Param("id")
+
+	type updateAccountStatusBody struct {
```
[the remark this answers is not in the document]
Comment (datphamcode295) [reply pkg/handler/project/project.go:101]: thank you, it's done

File: pkg/handler/project/project.go (hunk 3 above)
Comment (datphamcode295) [reply pkg/handler/project/project.go:119]: thank you, it's done

File: pkg/handler/project/project_test.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			w := httptest.NewRecorder()
+			ctx, _ := gin.CreateTestContext(w)
+			bodyReader := strings.NewReader(`
+			{
+				"status":"active"
+			}
+			`)
```
Comment (nguyennh4522) [inline pkg/handler/project/project_test.go:51]: the body must be get from `tests`, not declared in running phase

File: pkg/routes/v1.go [hunk 6]
```diff
@@ -38,4 +38,5 @@ func loadV1Routes(r *gin.Engine, h *handler.Handler, cfg *config.Config) {
 
 	// projects
 	v1.GET("/projects", h.Project.List)
+	v1.POST("/projects/:id/status", h.Project.UpdateProjectStatus)
```
Comment (nguyennh4522) [inline pkg/routes/v1.go:41]: should be `PUT` for update something

File: pkg/handler/project/project_test.go (hunk 5 above)
Comment (nguyennh4522) [reply pkg/handler/project/project_test.go:51]: example `https://github.com/dwarvesf/fortress-api/pull/29/files#diff-6d6a317eb1b0c03494708990c7cf6d81dbaad29a27adeed3b7d717b7cd6419dc`

Comment (datphamcode295) [reply pkg/handler/project/project_test.go]: thank you, it's done

File: pkg/routes/v1.go (hunk 6 above)
Comment (datphamcode295) [reply pkg/routes/v1.go]: thank you, it's done

File: pkg/handler/project/project_test.go [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+			expRespRaw, err := ioutil.ReadFile(tt.wantResponsePath)
+			require.NoError(t, err)
+
+			var actualData view.UpdateEmployeeStatusResponse
+			var expectedData view.UpdateEmployeeStatusResponse
+			err = json.Unmarshal(w.Body.Bytes(), &actualData)
+			err = json.Unmarshal(expRespRaw, &expectedData)
+
+			actualData.Data.UpdatedAt = nil
+			expectedData.Data.UpdatedAt = nil
```
Comment (nguyennh4522) [inline pkg/handler/project/project_test.go:71]: you can use `utils.RemoveFieldInResponse` here

File: pkg/handler/project/project_test.go [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+			`,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			w := httptest.NewRecorder()
+			ctx, _ := gin.CreateTestContext(w)
+			bodyReader := strings.NewReader(tt.request)
+			ctx.Params = gin.Params{gin.Param{Key: "id", Value: "8dc3be2e-19a4-4942-8a79-56db391a0b15"}}
```
Comment (nguyennh4522) [inline pkg/handler/project/project_test.go:54]: don't use fixed value in running phase

File: pkg/handler/project/project_test.go [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+	loggerMock := logger.NewLogrusLogger()
+	serviceMock := service.New(&cfg)
+	storeMock := store.New(&cfg)
+
+	tests := []struct {
+		name             string
+		wantCode         int
+		wantErr          error
+		wantResponsePath string
+		request          string
```
Comment (nguyennh4522) [inline pkg/handler/project/project_test.go:34]:
```suggestion
		request          updateAccountStatusBody
```

File: pkg/handler/project/project_test.go [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+			name:             "ok_update_project_status",
+			wantCode:         200,
+			wantErr:          nil,
+			wantResponsePath: "testdata/update_project_status/200.json",
+			request: `
+			{
+				"status":"active"
+			}
+			`,
+		},
```
Comment (nguyennh4522) [inline pkg/handler/project/project_test.go:46]: add more test cases (success/failure)

**Comment by namnhce** [comment]
Update screenshot

**Comment by nguyennh4522** [comment]
Update PR title

File: pkg/handler/project/project_test.go (hunk 8 above)
Comment (datphamcode295) [reply pkg/handler/project/project_test.go]: I moved it to tests struct

File: pkg/handler/project/project_test.go (hunk 9 above)
Comment (datphamcode295) [reply pkg/handler/project/project_test.go]: thanks you, it's done

File: pkg/handler/project/project_test.go (hunk 10 above)
Comment (datphamcode295) [reply pkg/handler/project/project_test.go:66]: thanks you, it's done

File: pkg/handler/project/project.go [hunk 11]
```diff
@@ -78,3 +78,56 @@ func (h *handler) List(c *gin.Context) {
...
+// @Accept  json
+// @Produce  json
+// @Param id path string true "Project ID"
+// @Param status body model.ProjectStatus true "Project Status"
+// @Param Authorization header string true "jwt token"
+// @Success 200 {object} view.UpdateProjectStatusResponse
+// @Failure 400 {object} view.ErrorResponse
+// @Failure 404 {object} view.ErrorResponse
+// @Failure 500 {object} view.ErrorResponse
+// @Router /projects/{id}/status [post]
```
Comment (namnhce) [inline pkg/handler/project/project.go:95]: update HTTP method
```suggestion
// @Router /projects/{id}/status [put]
```

**Review by namnhce — APPROVED** [review]

**Pull request merged**